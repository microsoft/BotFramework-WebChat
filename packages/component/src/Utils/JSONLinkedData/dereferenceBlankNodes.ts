import deepFreeze from 'deep-freeze-strict';

import { type BlankNode } from './BlankNode';
import getSafeOwnPropertyNames from './getSafeOwnPropertyNames';
import isBlankNode from './isBlankNode';
import isUnconnectedBlankNode from './isUnconnectedBlankNode';
import visitOnce from './visitOnce';

function dereferenceBlankNodesInline(objects: object[]): void {
  const blankNodeIdMap = new Map<string, BlankNode>();

  const objectsToVisit1: any[] = [objects];
  const shouldVisit1 = visitOnce<object>();

  while (objectsToVisit1.length) {
    const object = objectsToVisit1.shift();

    if (!object) {
      continue;
    }

    const indices = getSafeOwnPropertyNames(object);

    for (const index of indices) {
      // eslint-disable-next-line security/detect-object-injection
      const value = object[index];

      if (isBlankNode(value) && !isUnconnectedBlankNode(value)) {
        blankNodeIdMap.set(value['@id'], value);
      }

      shouldVisit1(value) && objectsToVisit1.push(value);
    }
  }

  const objectsToVisit2: unknown[] = [objects];
  const shouldVisit2 = visitOnce<object>();

  while (objectsToVisit2.length) {
    const object = objectsToVisit2.shift();

    if (!object) {
      continue;
    }

    const indices = getSafeOwnPropertyNames(object);

    for (const index of indices) {
      // eslint-disable-next-line security/detect-object-injection
      const value = object[index];

      if (isBlankNode(value) && isUnconnectedBlankNode(value)) {
        const blankNode = blankNodeIdMap.get(value['@id']);

        if (blankNode) {
          // eslint-disable-next-line security/detect-object-injection
          object[index] = blankNode;
        }
      } else {
        shouldVisit2(value) && objectsToVisit2.push(value);
      }
    }
  }
}

/**
 * Dereferences all unconnected blank nodes to their corresponding blank node. This is done by replacing all unconnected blank nodes in a graph and purposefully introduce cyclic dependencies to help querying the graph.
 *
 * This function will always return a new instance of all objects in the graph.
 *
 * This function assumes the graph conforms to JSON-LD, notably:
 *
 * - For nodes that share the same blank node identifier, one of them should be connected and the other must be unconnected
 * - If none of them are connected node, these unconnected blank node will not be replaced
 *
 * @see https://json-ld.github.io/json-ld.org/spec/latest/json-ld/#data-model-overview
 * @param graph A list of nodes in the graph.
 * @returns A structured clone of graph with unconnected blank nodes replaced by their corresponding blank node.
 */
export default function dereferenceBlankNodes<T extends object>(graph: T[]): readonly T[] {
  const nextObjects = structuredClone(graph);

  dereferenceBlankNodesInline(nextObjects);

  return deepFreeze(nextObjects);
}
