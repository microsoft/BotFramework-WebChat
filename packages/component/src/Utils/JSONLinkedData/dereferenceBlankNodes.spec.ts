import dereferenceBlankNodes from './dereferenceBlankNodes';

test('should leave non-blank node as-is', () => {
  const object = {
    '@type': 'Thing',
    first: { '@type': 'Thing' },
    second: { '@type': 'Thing', text: 'Hello, World!' },
    third: { text: 'Aloha!' }
  };

  const input = [object];
  const output = dereferenceBlankNodes(input);

  expect(output).not.toBe(input);

  const [nextObject] = output as any;

  expect(nextObject).not.toBe(object);

  expect(nextObject).toEqual({
    '@type': 'Thing',
    first: { '@type': 'Thing' },
    second: { '@type': 'Thing', text: 'Hello, World!' },
    third: { text: 'Aloha!' }
  });

  expect(Object.isFrozen(nextObject)).toBe(true);
  expect(Object.isFrozen(nextObject.first)).toBe(true);
  expect(Object.isFrozen(nextObject.second)).toBe(true);
  expect(Object.isFrozen(nextObject.third)).toBe(true);
});

test('should dereference unconnected blank nodes in cyclic fashion', () => {
  const object = {
    '@type': 'Book',
    author: {
      '@type': 'Person',
      '@id': '_:isaac',
      name: 'Isaac Stevens',
      worksFor: { '@type': 'Organization', '@id': '_:uw' }
    },
    name: 'Report of Explorations for a Route for the Pacific Railroad near the 47th and 49th Parallels of North Latitude, from St. Paul, Minnesota, to Puget Sound',
    sourceOrganization: {
      '@type': 'Organization',
      '@id': '_:uw',
      name: 'University of Washington',
      founder: { '@type': 'Person', '@id': '_:isaac' }
    }
  };

  const input = [object];
  const output = dereferenceBlankNodes(input);

  expect(output).not.toBe(input);

  const [nextObject] = output as any;

  expect(nextObject).toMatchSnapshot();

  expect(nextObject).not.toBe(object);
  expect(nextObject.author).not.toBe(object.author);
  expect(nextObject.sourceOrganization).not.toBe(object.sourceOrganization);

  expect(nextObject.author.worksFor).toBe(nextObject.sourceOrganization);
  expect(nextObject.sourceOrganization.founder).toBe(nextObject.author);

  expect(nextObject.author.name).toBe('Isaac Stevens');
  expect(nextObject.author.worksFor.name).toBe('University of Washington');
  expect(nextObject.sourceOrganization.name).toBe('University of Washington');
  expect(nextObject.sourceOrganization.founder.name).toBe('Isaac Stevens');

  expect(Object.isFrozen(nextObject)).toBe(true);
  expect(Object.isFrozen(nextObject.author)).toBe(true);
  expect(Object.isFrozen(nextObject.sourceOrganization)).toBe(true);
});

test('should dereference unconnected blank nodes in array in cyclic fashion', () => {
  const object = {
    '@type': 'Book',
    author: { '@type': 'Person', name: 'Edmond S. Meany' },
    about: [
      {
        '@type': 'Person',
        '@id': '_:isaac',
        name: 'Isaac Stevens',
        worksFor: {
          '@type': 'Organization',
          '@id': '_:uw',
          name: 'University of Washington',
          founder: { '@type': 'Person', '@id': '_:isaac' }
        }
      },
      {
        '@type': 'Organization',
        '@id': '_:uw'
      }
    ],
    name: 'Governors of Washington: territorial and state'
  };

  const input = [object];
  const output = dereferenceBlankNodes(input);

  expect(output).not.toBe(input);

  const [nextObject] = output as any;

  expect(nextObject).not.toBe(object);
  expect(nextObject).toMatchSnapshot();

  expect(nextObject.about[0].worksFor).toBe(nextObject.about[1]);
  expect(nextObject.about[1].founder).toBe(nextObject.about[0]);

  expect(nextObject.about[0].name).toBe('Isaac Stevens');
  expect(nextObject.about[0].worksFor.name).toBe('University of Washington');
  expect(nextObject.about[1].name).toBe('University of Washington');
  expect(nextObject.about[1].founder.name).toBe('Isaac Stevens');
});

test('should dereference unconnected blank nodes in the graph', () => {
  const input = [
    {
      '@type': 'Person',
      '@id': '_:isaac',
      name: 'Isaac Stevens',
      worksFor: {
        '@type': 'Organization',
        '@id': '_:uw',
        name: 'University of Washington',
        founder: { '@type': 'Person', '@id': '_:isaac' }
      }
    },
    {
      '@type': 'Organization',
      '@id': '_:uw'
    }
  ];

  const output = dereferenceBlankNodes(input) as any;

  expect(output).not.toBe(input);
  expect(output).toMatchSnapshot();

  expect(output[0].worksFor).toBe(output[1]);
  expect(output[1].founder).toBe(output[0]);

  expect(output[0].name).toBe('Isaac Stevens');
  expect(output[0].worksFor.name).toBe('University of Washington');
  expect(output[1].name).toBe('University of Washington');
  expect(output[1].founder.name).toBe('Isaac Stevens');

  expect(Object.isFrozen(output)).toBe(true);
  expect(Object.isFrozen(output[0])).toBe(true);
  expect(Object.isFrozen(output[1])).toBe(true);
});

test('should not dereference unconnectable blank nodes', () => {
  const object = { first: { '@id': '_:a' } };

  const [nextObject] = dereferenceBlankNodes([object]);

  expect(nextObject).toEqual({ first: { '@id': '_:a' } });

  expect(Object.isFrozen(nextObject)).toBe(true);
});

test('should not dereference unconnectable blank nodes in an array', () => {
  const object = { first: [{ '@id': '_:a' }] };

  const [nextObject] = dereferenceBlankNodes([object]);

  expect(nextObject).toEqual({ first: [{ '@id': '_:a' }] });

  expect(Object.isFrozen(nextObject)).toBe(true);
});
