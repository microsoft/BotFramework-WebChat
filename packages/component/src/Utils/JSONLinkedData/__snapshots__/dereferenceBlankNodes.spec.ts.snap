// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should dereference unconnected blank nodes in array in cyclic fashion 1`] = `
Object {
  "@type": "Book",
  "about": Array [
    Object {
      "@id": "_:isaac",
      "@type": "Person",
      "name": "Isaac Stevens",
      "worksFor": Object {
        "@id": "_:uw",
        "@type": "Organization",
        "founder": [Circular],
        "name": "University of Washington",
      },
    },
    Object {
      "@id": "_:uw",
      "@type": "Organization",
      "founder": Object {
        "@id": "_:isaac",
        "@type": "Person",
        "name": "Isaac Stevens",
        "worksFor": [Circular],
      },
      "name": "University of Washington",
    },
  ],
  "author": Object {
    "@type": "Person",
    "name": "Edmond S. Meany",
  },
  "name": "Governors of Washington: territorial and state",
}
`;

exports[`should dereference unconnected blank nodes in cyclic fashion 1`] = `
Object {
  "@type": "Book",
  "author": Object {
    "@id": "_:isaac",
    "@type": "Person",
    "name": "Isaac Stevens",
    "worksFor": Object {
      "@id": "_:uw",
      "@type": "Organization",
      "founder": [Circular],
      "name": "University of Washington",
    },
  },
  "name": "Report of Explorations for a Route for the Pacific Railroad near the 47th and 49th Parallels of North Latitude, from St. Paul, Minnesota, to Puget Sound",
  "sourceOrganization": Object {
    "@id": "_:uw",
    "@type": "Organization",
    "founder": Object {
      "@id": "_:isaac",
      "@type": "Person",
      "name": "Isaac Stevens",
      "worksFor": [Circular],
    },
    "name": "University of Washington",
  },
}
`;

exports[`should dereference unconnected blank nodes in the graph 1`] = `
Array [
  Object {
    "@id": "_:isaac",
    "@type": "Person",
    "name": "Isaac Stevens",
    "worksFor": Object {
      "@id": "_:uw",
      "@type": "Organization",
      "founder": [Circular],
      "name": "University of Washington",
    },
  },
  Object {
    "@id": "_:uw",
    "@type": "Organization",
    "founder": Object {
      "@id": "_:isaac",
      "@type": "Person",
      "name": "Isaac Stevens",
      "worksFor": [Circular],
    },
    "name": "University of Washington",
  },
]
`;
